### DISCUSSION TITLE
How can I disable arbitrary properties?

### DISCUSSION THREAD

**Discussion by LZL0** [tailwindlabs/tailwindcss#10933] [General]
I am having some issues with a legacy website's gulp configuration. For the staging server, I used to generate all the styles to make development easier. Since arbitrary properties, such as `w-[100px]`, were added gulp fails to generate the stylesheet.

How can I disable arbitrary properties? Could you guys please point me in the right direction?

**Comment by wongjn** [comment]
Just to be clear, do you mean arbitrary **properties**, like `[width:100px]` or arbitrary **values**, like `w-[100px]`? Either way, you can:

- Use the [`blocklist`](https://tailwindcss.com/docs/content-configuration#discarding-classes) to discard would-be rules for classes individually.
- Replace the [core plugins](https://github.com/tailwindlabs/tailwindcss/blob/master/src/corePlugins.js) that use [`matchUtilities()`](https://tailwindcss.com/docs/plugins#dynamic-utilities) and replace their registrations with [`addUtilities()`](https://tailwindcss.com/docs/plugins#static-utilities).
- Use a PostCSS plugin that runs after Tailwind that matches and discards rules based on a pattern of their selector.

**Reply by LZL0** [reply]
Thanks for the reply.
I would like to disable both of those. 
I used to run this selector, I presume this is what I need to update, provided that there is no other clear way to disable those two features in the config:

```
safelist: [
        {
            pattern: /./,
            variants: [
                "hover",
                "focus",
                "print",
                "sm",
                "md",
                "lg",
                "xl",
                "2xl",
            ],
        },
    ],
```

I am going to update this thread as I collect more info.

**Reply by paulgv** [reply]
We'd also be interested in a configuration option to disable arbitrary values 👍

**Reply by MPJHorner** [reply]
👍

**Comment by scr2em** [comment]
This is still a valid concern, I want to limit our developers to the design tokens

**Comment by uclaeamsavino** [comment]
We just ran into this problem. A dev has h-[64] (which should be h-[64px]) in one of her className properties:
```<div className="fixed top-12 left-0 right-0 z-40 h-[64] py-3 px-4">```

On her local, the height displays correctly as 64px:
<img width="388" height="53" alt="image" src="https://github.com/user-attachments/assets/0a45f24b-e62f-45e0-892a-9dc3d0dda744" />

However on our AWS Amplify-hosted nextjs server, h-[64] fails and the style is blocked:
<img width="379" height="63" alt="image" src="https://github.com/user-attachments/assets/7c9d19df-f558-4ec2-a32e-9919a0fffc1f" />

I've spent the last couple hours googling and hallucinating with ChatGPT and it seems like there's no simple fix for this? How can we know that what we see on our local is correct if it fails on the server? I even got the eslint tailwind plugin working and it found a bunch of other invalid tailwind class names, but it didn't trigger on h-[64]. 

This is a real problem if we can't get local to fail in the same way the server would fail. This time it was pretty obvious. But I could see hours of head-scratching in the future trying to figure out why the server is failing.

**Reply by rozsazoltan** [reply]
No. This is correct. https://play.tailwindcss.com/HgrSs3py8r - See generated CSS > Utilities > I think it's impossible (or if you think it can be reproduced in your local version, then share a corresponding reproduction on [play.tailwindcss.com](https://play.tailwindcss.com/)
 or as a GitHub repository.) for it to be 64px locally without external intervention.

### How to work by default

> |Class|Styles|
> |---|---|
> |`h-[<value>]`|`height: <value>;`|
> 
> * https://tailwindcss.com/docs/height

Following the official syntax, `h-[64]` will always correctly result in `height: 64;`; there's no automatic conversion to pixels, as you mentioned.

Furthermore, ChatGPT isn't suitable for finding this kind of issue, as it doesn't contain enough up-to-date information about v4 or similar bugs.

Final solution: use a valid value: `h-[64px]` instead of `h-[64]`

### Just interesting functional utility handling

But you can define a rule that handles the [integer] arbitrary value in a calculation, resulting in pixels:

<img width="409" height="413" alt="image" src="https://github.com/user-attachments/assets/777d0637-7fb2-4ee4-9841-cbf7c8729511" />
<img width="419" height="408" alt="image" src="https://github.com/user-attachments/assets/c16c5ecd-228a-4930-9112-e21a4a63fc16" />

Unfortunately, since the original also includes support for integer arbitrary values, this creates a duplication - but yours will take precedence. See:
* https://github.com/tailwindlabs/tailwindcss/discussions/18944

**Reply by uclaeamsavino** [reply]
Thank you for the well-thought-out reply. It's good to know that h-[64] should always fail and this isn't the expected behavior of tailwind in dev mode. 

However, something is causing h-[64] to not fail for us locally:
<img width="388" height="53" alt="image" src="https://github.com/user-attachments/assets/a201f15f-16e5-4eb4-bd8a-cc9f704dce4d" />
I will investigate and see if I can find the source of this, or put together a demo.

**Reply by uclaeamsavino** [reply]
Okay, I have more clues about what's happening. When I run ```next build``` locally, it creates a .next folder. When I search that folder for h-\[64, I only see this file where the class name is not corrected:
<img width="673" height="224" alt="image" src="https://github.com/user-attachments/assets/bf6890d5-fac5-40c1-b2cd-95e75967941a" />

If I then run ```next start``` locally to run the production build, the class name is not corrected and I see the CSS error.

However, if I run ```next dev --turbopack``` locally, and then hit the site, the JIT compiler creates a new globals.css file, where apparently it "fixes" the broken class name:
<img width="648" height="121" alt="image" src="https://github.com/user-attachments/assets/2a54ae6b-4341-424d-9795-9b3d6728a4af" />

So this is definitely the source of our issue. For some reason the class name is getting corrected in dev mode only. I'd love to just turn off whatever is doing that fixing. Dev should be stricter or as strict as production, never less strict.

**Reply by uclaeamsavino** [reply]
Update: Removing ```--turbopack``` from the next dev call causes the class to not get fixed. So it's something turbopack is doing or calling. We are on next 15.5.4 which uses webpack by default. The investigation continues.

I fully realize that ChatGPT might be hallucinating all of this. But just for reference, this is what it says:
<img width="806" height="914" alt="image" src="https://github.com/user-attachments/assets/c8ead8ae-29c2-40a7-ab9d-6f43d8fe21bd" />

**Reply by rozsazoltan** [reply]
@uclaeamsavino Aha, I took a look.

> Turbopack API Reference - Lightning CSS: In Use

It seems that [Turbopack runs the result through LightningCSS in the background](https://nextjs.org/docs/app/api-reference/turbopack#css-and-styling). That's why, in development mode, Lightning CSS "magically" converts the invalid value `64` into `64px`.

> Under the hood Tailwind uses [Lightning CSS](https://lightningcss.dev/) to process nested CSS.

TailwindCSS does this as well, [but only during the build process](https://github.com/tailwindlabs/tailwindcss/blob/dc6a3ce349a082e6c437288e52d0eea9a9ea4ff2/packages/%40tailwindcss-node/src/optimize.ts#L33-L56) - whereas Turbopack does it even during development.

 **I think this is bad practice, and the simple solution (use `[64px]`, `[4rem]`, etc.) is to just avoid using invalid length values (e.g. `[64]`).**

As for why it doesn't become `64px` again in the build output, I haven't figured that out yet. In theory, with the TailwindCSS configuration, Lightning CSS should also correct the invalid length values during the build.
* [Lightning CSS playground](https://lightningcss.dev/playground/index.html#%7B%22minify%22%3Afalse%2C%22customMedia%22%3Atrue%2C%22cssModules%22%3Afalse%2C%22analyzeDependencies%22%3Afalse%2C%22targets%22%3A%7B%22chrome%22%3A7274496%2C%22firefox%22%3A8388608%2C%22safari%22%3A1048576%2C%22ios_saf%22%3A1048576%7D%2C%22include%22%3A449%2C%22exclude%22%3A1572868%2C%22source%22%3A%22.h-%5C%5C%5B64%5C%5C%5D%20%7B%5Cn%20%20height%3A%2064%3B%5Cn%7D%22%2C%22visitorEnabled%22%3Afalse%2C%22visitor%22%3A%22%7B%5Cn%20%20Color(color)%20%7B%5Cn%20%20%20%20if%20(color.type%20%3D%3D%3D%20'rgb')%20%7B%5Cn%20%20%20%20%20%20color.g%20%3D%200%3B%5Cn%20%20%20%20%20%20return%20color%3B%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%22%2C%22unusedSymbols%22%3A%5B%5D%2C%22version%22%3A%22local%22%7D)

**Reply by uclaeamsavino** [reply]
Well as it turns out, we weren't using the --turbopack option for ```next build```. We were only using it for ```next dev```. And the mystery is solved as to why we weren't seeing this issue in our server running the build, but were seeing it locally running dev. 

I inherited this project and that's how it was set up. I assumed --turbopack was just something for dev use. Now I just have to decide if I want to add ```--turbooack``` for both options (nextjs 15.5.x) or for neither option. Do you have any strong opinions on that? I guess getting used to turbopack now might make more sense for when we upgrade to nextjs 16, where it's the default.

Thank you so much for jumping in on this and helping me debug our setup. I really appreciate it.
